tests/integration/test_total_max_threads/test.py

```diff
@@ -0,0 +1,49 @@
+import pytest
+from helpers.cluster import ClickHouseCluster
+
+cluster = ClickHouseCluster(__file__)
+node1 = cluster.add_instance(
+    "node1",
+    main_configs=["configs/config_default.xml"],
+    user_configs=["configs/users.xml"],
+)
+node2 = cluster.add_instance(
+    "node2",
+    main_configs=["configs/config_defined.xml"],
+    user_configs=["configs/users.xml"],
+)
+
+
+@pytest.fixture(scope="module")
+def started_cluster():
+    try:
+        cluster.start()
+        yield cluster
+    finally:
+        cluster.shutdown()
+
+
+def test_total_max_threads_default(started_cluster):
+    node1.query(
+        "SELECT count(*) FROM numbers_mt(10000000)", query_id="test_total_max_threads_1"
+    )
+    node1.query("SYSTEM FLUSH LOGS")
+    assert (
+        node1.query(
+            "select length(thread_ids) from system.query_log where current_database = currentDatabase() and type = 'QueryFinish' and query_id = 'test_total_max_threads_1'"
+        )
+        == "102\n"
+    )
+
+
+def test_total_max_threads_defined(started_cluster):
+    node2.query(
+        "SELECT count(*) FROM numbers_mt(10000000)", query_id="test_total_max_threads_2"
+    )
+    node2.query("SYSTEM FLUSH LOGS")
+    assert (
+        node2.query(
+            "select length(thread_ids) from system.query_log where current_database = currentDatabase() and type = 'QueryFinish' and query_id = 'test_total_max_threads_2'"
+        )
+        == "51\n"
+    )
```

Comment:
Test does not check that limit will be decreased when the threshold will be reached. This case sounds more major then the default value.

Reply:
@azat Could you please explain more in details what do you mean? Because in fact this test test_total_max_threads_defined shows that when the query requires 102 threads to execute and we have total_max_threads=50 then finally the query is executed in 52 threads. It's less than 102. So it's OK. Please explain your point of view with sample numbers.

Reply:
Yes, it's an important point that in any case the query still has one thread so the in even worst case the query will be allowed to execute but in one thread. And for high CPU load conditions it's in fact the best case. It's like max_threads=1. 
Now I am working on the documentation text to clearly explain this point. 

But my question was about the integration test test_total_max_threads_defined. Do you think it's OK or how should I improve it? Because in my option the test it OK.

Reply:
@azat So do you mean I should simply write a new (3rd) test for the when 1 thread will be recommended by system and finally the query will be executed in 2 threads? 
Then we could verify it by
length(thread_ids) = 2 
Do you mean that?

Reply:
@azat I tried this code:

```
# total_max_threads = 10 in config.
def test_total_max_threads_reached(started_cluster):
    node4.query(
        "SELECT count(*) FROM numbers_mt(10000000)  settings max_threads=5"
    )
    node4.query(
        "SELECT count(*) FROM numbers_mt(10000000)  settings max_threads=5"
    )
    node4.query(
        "SELECT count(*) FROM numbers_mt(10000000)  settings max_threads=5", query_id="test_total_max_threads_4"
    )    
    node4.query("SYSTEM FLUSH LOGS")
    assert (
        node4.query(
            "select length(thread_ids) from system.query_log where current_database = currentDatabase() and type = 'QueryFinish' and query_id = 'test_total_max_threads_4'"
        )
        == "2\n"
    )

```
The problem is that search query is executed in sync mode. So  length(thread_ids)=7 instead of 2.

Probably I need use use some async features of CH or Python to simulate it. Do you have any suggestions?

Reply:
You can execute queries in threads, see https://github.com/ClickHouse/ClickHouse/blob/a53cfa9fcad121830db211a5831c3eafc36fb1ac/tests/integration/test_storage_kafka/test.py#L1149-L1158 for example

Reply:
@azat OK. Thank you. I will try it. I also found the following interesting apply_async usage:
https://github.com/ClickHouse/ClickHouse/blob/a53cfa9fcad121830db211a5831c3eafc36fb1ac/tests/integration/test_quorum_inserts_parallel/test.py#L46

Reply:
@azat See the latest commit. I added the following test:

```
# note: total_max_threads = 10 in config_limit_reached.xml
def test_total_max_threads_limit_reached(started_cluster):
    logging.debug("ADQM: test begin")

    def thread_select():
        logging.debug("ADQM: started another thread")
        node4.query(
            "SELECT sleep(3) FROM numbers_mt(10000000) settings max_threads=100"
        )
        logging.debug("ADQM: finished another thread")

    another_thread = threading.Thread(target=thread_select)
    another_thread.start()

    time.sleep(0.5)
    logging.debug("ADQM: started main query")
    node4.query(
        "SELECT count(*) FROM numbers_mt(10000000) settings max_threads=5",
        query_id="test_total_max_threads_4",
    )
    logging.debug("ADQM: finished main query")
    another_thread.join()
    logging.debug("ADQM: logs: %s", node4.grep_in_log("ADQM"))
    node4.query("SYSTEM FLUSH LOGS")
    assert (
        node4.query(
            "select length(thread_ids) from system.query_log where current_database = currentDatabase() and type = 'QueryFinish' and query_id = 'test_total_max_threads_4'"
        )
        == "2\n"
    )
    logging.debug("ADQM: test end")
```

According to logs it seems like it works fine:

> 
> 2022.05.20 11:11:32.481001 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: max_threads: 100
> 2022.05.20 11:11:32.481003 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Recommended num threads: 100
> 2022.05.20 11:11:32.481006 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Total number of threads from config: 10
> 2022.05.20 11:11:32.481012 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Current total num threads: 1
> 2022.05.20 11:11:32.481015 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Total available threads: 9
> 2022.05.20 11:11:32.481017 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Recommended num threads: 9
> 2022.05.20 11:11:32.481020 [ 8 ] {631d3398-20b7-4c5f-ba52-1b3d7456a18b} <Debug> ADQM: Finally num threads: 9
> 2022.05.20 11:11:33.019071 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: maxParallelStreams: 5
> 2022.05.20 11:11:33.019077 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: max_threads: 5
> 2022.05.20 11:11:33.019080 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Recommended num threads: 5
> 2022.05.20 11:11:33.019083 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Total number of threads from config: 10
> 2022.05.20 11:11:33.019089 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Current total num threads: 12
> 2022.05.20 11:11:33.019093 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Total available threads: 0
> 2022.05.20 11:11:33.019096 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Recommended num threads: 0
> 
> 2022.05.20 11:11:33.019099 [ 9 ] {test_total_max_threads_4} <Debug> ADQM: Finally num threads: 1

@azat Could you please comment it? Is it OK? The idea was to make one long thread working in a separate thread which consumes all query processing threads (it reaches the limit). And at the same time run our control query to verify that our algorithm gives only 1 thread to run to this query. It seems like it's working. I was in doubt about running the the long query in the same node4. But in seems like it's working: both queries runs in parallel. What do you think?

Reply:
Yes LGTM.